picard/script.py

```diff
@@ -1201,3 +1202,32 @@ def func_slice(parser, multi, start_index, end_index, separator=MULTI_VALUED_JOI
         return separator.join(multi_var[start:end])
     except IndexError:
         return ''
+
+
+@script_function()
+def func_datetime(parser, format=None, preset=None):
+    """Return the current date and time as a string.
+
+    Arguments:
+        parser: The ScriptParser object used to parse the script.
+        format: A string or the ScriptVariable/Function that evaluates to the
+            string used to format the output.  Default is '%Y-%m-%d %H:%M:%S'
+            if blank.  Uses strftime() format.
+        preset: A string or the ScriptVariable/Function that evaluates to the
+            string used to set the date and time value to be used.  Must be in
+            the format '%Y-%m-%d %H:%M:%S'.
+
+    Returns:
+        Returns he current date and time as a string.
+    """
+    # Handle case where format evaluates to ''
+    if not format:
+        format = '%Y-%m-%d %H:%M:%S'
+    if preset:
+        try:
+            time_now = datetime.datetime.strptime(preset, '%Y-%m-%d %H:%M:%S')
+        except ValueError:
+            time_now = datetime.datetime.now()
+    else:
+        time_now = datetime.datetime.now()
```

Comment:
I'm not sure to understand this new code.
Also this is repeated, it should be possible to factorise.
The string `'%Y-%m-%d %H:%M:%S'` is still repeated.

Does this code actually provide timezone information (`%z` and '%Z') ?

Reply:
The short answer is no, this does not provide the timezone information.  I am still working on trying to do that, but wanted to leave the code in a (semi) working state while I looked for solutions.  The problem is that Python 3.5 does not allow applying the `astimezone()` method to a niave `datetime` object, and I have yet to come up with a reliable means of getting the local `tzinfo` to use when creating the `datetime` object using `strptime`.

Reply:
I found a way to get the local timezone to set it when creating the `datetime` objects.  This should now provide the timezone information to allow the `%Z` and `%z` format specifiers to work properly.